Automattic/jetpack
# AI Assistant: Stop event stream request when block gets deleted

Fixes #31390.

## Proposed changes:

* Add a cleanup function to close the event stream source when the component gets unmounted, so the block stops the completion event stream when it's deleted

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






* Go to the block editor
* Open your browser console to inspect the debug messages of the block (remember to enable it)
* Add an AI Assistant block
* Ask for some completion that may take a few seconds to complete (for example: "give me 5 long paragraphs about the history of software development")
* Wait for the completion to start, then go the three-dot menu of the block and click "Remove AI Assistant (Experimen...":

<img width="644" alt="Screen Shot 2023-06-19 at 19 40 59" src="https://github.com/Automattic/jetpack/assets/6760046/f7f3ab26-257d-4f42-88c2-fe5e9b5a516e">

* Watch the debug logs on the console
* **Before the fix**, the completion will process until the end, and we can see the debug messages getting logged even after the block is gone (what is not expected):

https://github.com/Automattic/jetpack/assets/77539/628f786f-602e-4c5c-bec6-517cdb771f76

* **After the fix**, the completion will stop immediately (a `Cleaning things up...` debug message will be the last thing):

<img width="610" alt="Screen Shot 2023-06-19 at 19 45 02" src="https://github.com/Automattic/jetpack/assets/6760046/cccf81f5-ee60-499b-8ecf-bee1f3d1b030">
